File: Algorithms/189. Rotate Array/php/solution.cyclic_replacement.php
<?php
class Solution
{

    /**
     * @param Integer[] $nums
     * @param Integer $k
     * @return NULL
     */
    function rotate(&$nums, $k)
    {
        $n = count($nums);
        $k %= $n;

        $start = $count = 0;
        while ($count < $n) {
            $current = $start;
            $prev = $nums[$start];

            do {
                $nextIdx = ($current + $k) % $n;
                $temp = $nums[$nextIdx];
                $nums[$nextIdx] = $prev;
                $prev = $temp;
                $current = $nextIdx;
                $count++;
            } while ($current != $start);
            $start++;
        }
    }
}

File: Algorithms/189. Rotate Array/golang/solution.cyclic_replacement.go
func rotate(nums []int, k int) {
	var n int = len(nums)
	k = k % n

	var start int = 0
	var count int = 0
	for count < n {
		current, prev := start, nums[start]
		for current != start {
			nextIdx := (current + k) % n
			nums[nextIdx], prev = prev, nums[nextIdx]
			current = nextIdx
			count++
		}

		start++
	}
}